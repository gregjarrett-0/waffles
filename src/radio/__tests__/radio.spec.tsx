import { useRef, useEffect } from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Radio } from '../index';

const MOCKED_ID = '123abC';

jest.mock('nanoid', () => {
  return {
    nanoid: () => MOCKED_ID,
  };
});

function TestRefRadio() {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <Radio ref={inputRef} onChange={jest.fn()}>
      Test label
    </Radio>
  );
}

describe('Radio', () => {
  it('renders label and input', () => {
    const { getByLabelText, getByText } = render(
      <Radio onChange={jest.fn()}>Taylor Swift</Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    const label = getByText('Taylor Swift');

    expect(input).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it('handles click event correctly', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Radio onChange={handleClick}>Taylor Swift</Radio>,
    );

    const label = getByText('Taylor Swift');
    fireEvent.click(label);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('handles focus correctly', () => {
    const handleFocus = jest.fn();
    const { getByLabelText } = render(
      <Radio onFocus={handleFocus} onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    fireEvent.focus(input);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('sets the data attribute on the radio', () => {
    const { getByTestId } = render(
      <Radio data-testid="taylors-radio" onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const input = getByTestId('taylors-radio');

    expect(input).toBeInTheDocument();
  });

  it('label and input are associated by the same ID', () => {
    const { getByLabelText, getByText } = render(
      <Radio onChange={jest.fn()}>Taylor Swift</Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    const label = getByText('Taylor Swift');

    expect(input).toHaveAttribute('id', `radio-${MOCKED_ID}`);
    expect(label).toHaveAttribute('for', `radio-${MOCKED_ID}`);
  });

  it('sets correct aria attributes when has error', () => {
    const { getByLabelText } = render(
      <Radio error onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });

    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  it('renders the disabled radio', () => {
    const { getByLabelText } = render(
      <Radio disabled onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });

    expect(input).toBeDisabled();
  });

  it('accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefRadio />);

    const input = container.querySelector('input');

    expect(input).toHaveFocus();
  });

  it('renders basic snapshot', () => {
    const { container } = render(
      <Radio onChange={jest.fn()}>Taylor Swift</Radio>,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot of radio without label', () => {
    const { container } = render(
      <Radio aria-label="Taylor Swift" onChange={jest.fn()} />,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot of checked state', () => {
    const { container } = render(
      <Radio checked onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot with error', () => {
    const { container } = render(
      <Radio error onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot of inverted', () => {
    const { container } = render(
      <Radio inverted onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot of disabled state', () => {
    const { container } = render(
      <Radio disabled onChange={jest.fn()}>
        Taylor Swift
      </Radio>,
    );

    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });

  it('renders snapshot of focused state', () => {
    const { container, getByLabelText } = render(
      <Radio onChange={jest.fn()}>Taylor Swift</Radio>,
    );

    const input = getByLabelText('Taylor Swift', { selector: 'input' });
    fireEvent.focus(input);
    const radio = container.firstChild;

    expect(radio).toMatchSnapshot();
  });
});
