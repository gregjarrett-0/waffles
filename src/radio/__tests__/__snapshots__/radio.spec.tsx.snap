// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders basic snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  />
  <label
    class="emotion-3"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;

exports[`Radio renders snapshot of checked state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #05192D;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: block;
  width: 8px;
  height: 8px;
  background-color: #009BD8;
  border-radius: 50%;
}

.emotion-4 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    checked=""
    class="emotion-1"
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
  </div>
  <label
    class="emotion-4"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;

exports[`Radio renders snapshot of disabled state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 0.6;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: default;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: default;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    disabled=""
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  />
  <label
    class="emotion-3"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;

exports[`Radio renders snapshot of focused state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-2::after {
  box-shadow: 0 0 0 2px #009BD8;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  />
  <label
    class="emotion-3"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;

exports[`Radio renders snapshot of inverted 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #213147;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #FFFFFF;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    class="emotion-1"
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  />
  <label
    class="emotion-3"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;

exports[`Radio renders snapshot of radio without label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="false"
    aria-label="Taylor Swift"
    class="emotion-1"
    type="radio"
  />
  <div
    class="emotion-2"
  />
</div>
`;

exports[`Radio renders snapshot with error 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  position: absolute;
  opacity: 0;
  width: 18px;
  height: 18px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 18px;
  height: 18px;
  border-radius: 50%;
  -webkit-transition: background-color 75ms ease-out,border-color 75ms ease-out;
  transition: background-color 75ms ease-out,border-color 75ms ease-out;
  pointer-events: none;
  background-color: #FFFFFF;
  border-width: 1px;
  border-style: solid;
  border-color: #848C92;
  border-color: #DD3400;
  box-shadow: 0 0 0 1px #DD3400;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -3px;
  border-radius: 50%;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 1px;
  padding-left: 8px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <input
    aria-invalid="true"
    class="emotion-1"
    id="radio-123abC"
    type="radio"
  />
  <div
    class="emotion-2"
  />
  <label
    class="emotion-3"
    for="radio-123abC"
  >
    Taylor Swift
  </label>
</div>
`;
